// @markwhen/parser@1.0.1/lib/dateRange/checkRecurrence.js
import { DateTime } from "luxon";
import { edtf_recurrenceMatchIndex, recurrence_recurrenceMatchIndex, } from "../regex.js";
import { RangeType } from "../Types.js";
import * as rrule from "@markwhen/rrule";
const { RRule } = rrule;
export const recurrenceDurationUnits = [
    "years",
    "months",
    "weeks",
    "weekdays",
    "days",
    "hours",
    "minutes",
    "seconds",
    "milliseconds",
];
function makeRecurrenceSerializable(recurrence) {
    const until = recurrence.until
        ? DateTime.fromJSDate(recurrence.until)?.toISO() ?? undefined
        : undefined;
    const dtstart = recurrence.dtstart
        ? DateTime.fromJSDate(recurrence.dtstart)?.toISO() ?? undefined
        : undefined;
    return { ...recurrence, until, dtstart };
}
export function toJsDates(recurrence) {
    const until = recurrence.until
        ? DateTime.fromISO(recurrence.until).toJSDate()
        : undefined;
    const dtstart = recurrence.dtstart
        ? DateTime.fromISO(recurrence.dtstart).toJSDate()
        : undefined;
    return {
        ...recurrence,
        until,
        dtstart,
    };
}
export const checkEdtfRecurrence = (line, i, lengthAtIndex, eventStartLineRegexMatch, context) => {
    const recurrenceMatch = eventStartLineRegexMatch[edtf_recurrenceMatchIndex];
    if (!recurrenceMatch) {
        return;
    }
    const recurrenceMatchTrimmed = recurrenceMatch.trim();
    const indexInString = eventStartLineRegexMatch[0].indexOf(recurrenceMatchTrimmed);
    const from = lengthAtIndex[i] + indexInString;
    const to = lengthAtIndex[i] + indexInString + recurrenceMatchTrimmed.length;
    const range = {
        type: RangeType.Recurrence,
        from,
        to,
        content: recurrenceMatch,
    };
    try {
        const recurrence = RRule.parseText(recurrenceMatch);
        return { recurrence: makeRecurrenceSerializable(recurrence), range };
    }
    catch (e) {
        context.parseMessages.push({
            type: "error",
            message: "Cannot parse recurrence",
            pos: [from, to],
        });
        return;
    }
};
export const checkRecurrence = (line, i, lengthAtIndex, eventStartLineRegexMatch, context) => {
    const recurrenceMatch = eventStartLineRegexMatch[recurrence_recurrenceMatchIndex];
    if (!recurrenceMatch) {
        return;
    }
    const recurrenceMatchTrimmed = recurrenceMatch.trim();
    const indexInString = eventStartLineRegexMatch[0].indexOf(recurrenceMatchTrimmed);
    const from = lengthAtIndex[i] + indexInString;
    const to = lengthAtIndex[i] + indexInString + recurrenceMatchTrimmed.length;
    const range = {
        type: RangeType.Recurrence,
        from,
        to,
        content: recurrenceMatch,
    };
    try {
        const recurrence = RRule.parseText(recurrenceMatch);
        return { recurrence: makeRecurrenceSerializable(recurrence), range };
    }
    catch (e) {
        context.parseMessages.push({
            type: "error",
            message: "Cannot parse recurrence",
            pos: [from, to],
        });
        return;
    }
};
