// @markwhen/parser@1.0.1/lib/incremental.js
import { ChangeSet, Text } from "@codemirror/state";
import { get, isEvent, isGroup, iter, iterateTreeFromPath, } from "./Types.js";
import { parse, parsePastHeader } from "./parse.js";
import { ParsingContext } from "./ParsingContext.js";
function touchesRanges(rangeA, [fromB, toB]) {
    let fromA;
    let toA;
    if (Array.isArray(rangeA)) {
        [fromA, toA] = rangeA;
    }
    else {
        fromA = rangeA.from;
        toA = rangeA.to;
    }
    return toA >= fromB && toB >= fromA;
}
const next = (it) => {
    let { done, value, } = it.next();
    return done ? { done } : { done, eventy: value.eventy, path: value.path };
};
function getPreviousEventPath(root, path) {
    let previousPath;
    const target = path.join(",");
    for (const { eventy, path: currentPath } of iter(root)) {
        if (!isEvent(eventy)) {
            continue;
        }
        if (currentPath.join(",") === target) {
            return previousPath ? [...previousPath] : undefined;
        }
        previousPath = currentPath;
    }
    return undefined;
}
function asArray(changes) {
    const c = [];
    changes.iterChanges((fromA, toA, fromB, toB, inserted) => {
        c.push({ fromA, toA, fromB, toB, inserted });
    });
    return c;
}
// skip this subtree and go to next sibling or up to next uncle
function skip(root, p) {
    let path = [...p];
    const last = path.at(-1);
    if (last === undefined) {
        return;
    }
    path = [...path.slice(0, -1), last + 1];
    while (!get(root, path)) {
        // We can't just pop because we don't want to just go up,
        // that would just put us at our parent's and liable
        // to reenter the node we're trying to skip.
        path.pop();
        const last = path.at(-1);
        if (last === undefined) {
            return;
        }
        path = [...path.slice(0, -1), last + 1];
    }
    return path;
}
function getCommonAncestor(arrays) {
    if (arrays.length === 0)
        return [];
    let prefix = [];
    // Use the first array as a base for comparison
    const firstArray = arrays[0];
    for (let i = 0; i < firstArray.length; i++) {
        const currentValue = firstArray[i];
        // Check if this value is present at the same position in all other arrays
        for (let j = 1; j < arrays.length; j++) {
            if (i >= arrays[j].length || arrays[j][i] !== currentValue) {
                return prefix;
            }
        }
        // If all arrays have the same value at position i, add it to the prefix
        prefix.push(currentValue);
    }
    return prefix;
}
function getFromAndTo(change, affected, preChange = true) {
    let min = Number.POSITIVE_INFINITY, max = -1;
    if (!affected.length) {
        change.iterChangedRanges((fromA, toA, fromB, toB) => {
            if (fromB < min) {
                min = fromB;
            }
            if (toB > max) {
                max = toB;
            }
        });
        return [min, max];
    }
    for (let i = 0; i < affected.length; i++) {
        const { from, to } = affected[i].textRanges.whole;
        if (from < min) {
            min = from;
        }
        if (to > max) {
            max = to;
        }
    }
    let newFrom = preChange ? change.mapPos(min) : min;
    let newTo = preChange ? change.mapPos(max) : max;
    if (preChange) {
        change.iterChangedRanges((fromA, toA, fromB, toB) => {
            if (fromB < newFrom) {
                newFrom = fromB;
            }
            if (toB > newTo) {
                newTo = toB;
            }
        });
    }
    return [newFrom, newTo];
}
function linesAndLengths(newText, change, affected, preChange = true) {
    const [newFrom, newTo] = getFromAndTo(change, affected, preChange);
    const lineFrom = newText.lineAt(newFrom);
    const lineTo = newText.lineAt(newTo);
    // Make sure we get complete lines
    const textSlice = newText.slice(lineFrom.from, newTo);
    const textLines = textSlice.toString().split("\n");
    const lines = textLines;
    const lengths = [];
    // Initialize with the starting position
    lengths.push(lineFrom.from);
    let runningLength = lineFrom.from;
    for (let i = 0; i < lines.length; i++) {
        runningLength += lines[i].length + (i === lines.length - 1 ? 0 : 1);
        lengths.push(runningLength);
    }
    return { from: lineFrom.number - 1, lines, lengths };
}
function getIds(eventies) {
    const ids = [];
    for (let i = 0; i < eventies.length; i++) {
        const e = eventies[i];
        if (isGroup(e)) {
            ids.push(...getIds(e.children));
        }
        else {
            if (e.id) {
                ids.push(e.id);
            }
        }
    }
    return ids;
}
function getTags(eventies) {
    const tags = [];
    for (let i = 0; i < eventies.length; i++) {
        const e = eventies[i];
        if (isGroup(e)) {
            tags.push(...e.tags);
            tags.push(...getTags(e.children));
        }
        else {
            if (e.tags) {
                tags.push(...e.tags);
            }
        }
    }
    return tags;
}
function graft({ changedRange, previousParse, done, eventy, path, eventyIterator, previousText, now, }) {
    const { events: root } = previousParse;
    const { fromA, toA, fromB, inserted } = changedRange;
    const n = () => {
        const iterate = next(eventyIterator);
        if (iterate.done) {
            done = true;
        }
        else {
            done = false;
            eventy = iterate.eventy;
            path = iterate.path;
        }
    };
    // If we're above the first event (aka the header area) abort
    if (toA < eventy.textRanges.whole.from ||
        fromA < eventy.textRanges.whole.from) {
        throw new Error("Can't reparse in header area");
    }
    let lastUnaffected = { eventy, path };
    let prior;
    let noneAffected = false;
    while (!done && !touchesRanges(eventy.textRanges.whole, [fromA, toA])) {
        if (prior) {
            lastUnaffected = prior;
        }
        prior = { eventy, path };
        if (isGroup(eventy)) {
            const newPath = skip(root, path);
            if (!newPath) {
                noneAffected = true;
                break;
            }
            eventyIterator = iterateTreeFromPath(root, newPath);
        }
        n();
    }
    const affected = [];
    if (!noneAffected) {
        // If we are editing the definition of an event or section, we might ablate it completely,
        // so we need to assume the prior event will be affected as well
        if (prior) {
            if (touchesRanges(eventy.textRanges.definition, [fromA, toA])) {
                affected.push(prior);
            }
            else {
                lastUnaffected = prior;
            }
        }
        // Now accumulate all the affected eventies
        do {
            affected.push({ eventy, path });
            n();
        } while (!done && touchesRanges(eventy.textRanges.whole, [fromA, toA]));
    }
    // If both a parent and children are affected, we can exclude the children
    // as we'll just reparse the whole parent node, children included.
    for (let i = 1; i < affected.length; i++) {
        const potentialChildPath = affected[i].path.join(",");
        const potentialParentPath = affected[i - 1].path.join(",");
        if (potentialChildPath.startsWith(potentialParentPath)) {
            affected.splice(i, 1);
            i--;
        }
    }
    const affectedEventies = affected.map(({ eventy }) => eventy);
    if (getIds(affectedEventies).length) {
        throw new Error("Won't reparse over ided events");
    }
    const _change = ChangeSet.of({ from: fromA, to: toA, insert: inserted }, previousText.length);
    const newText = previousText.replace(fromB, fromB + (toA - fromA), inserted);
    const withAffected = (affectedEventies, lastUnaffected) => {
        const { from, lines, lengths } = linesAndLengths(newText, _change, affectedEventies);
        const context = () => {
            const _context = new ParsingContext(now, previousParse.cache, (_c) => {
                if (_c.tail) {
                    return _c.tail;
                }
                if (lastUnaffected) {
                    if (isEvent(lastUnaffected)) {
                        return lastUnaffected;
                    }
                    throw new Error("Last unaffected eventy to relate to is not an event");
                }
            });
            _context.getById = (id) => {
                const path = previousParse.ids[id];
                if (!path) {
                    return;
                }
                return get(previousParse.events, path);
            };
            _context.header = previousParse.header;
            _context.ids = previousParse.ids;
            return _context;
        };
        const c = parsePastHeader(from, context(), Array(from).concat(lines), Array(from).concat(lengths));
        return {
            context: c,
            affectedFrom: lengths[0],
            affectedTo: lengths.at(-1),
        };
    };
    const withCommonAncestor = () => {
        const commonAncestor = getCommonAncestor(affected.map(({ path }) => path));
        if (commonAncestor.length) {
            const ancestor = get(root, commonAncestor);
            if (ancestor) {
                affected.splice(0, affected.length, {
                    path: commonAncestor,
                    eventy: ancestor,
                });
                return withAffected(affected.map(({ eventy }) => eventy), lastUnaffected.eventy);
            }
        }
    };
    let result;
    if (areSiblings(affected.map(({ path }) => path))) {
        result = withAffected(affected.map(({ eventy }) => eventy), lastUnaffected.eventy);
    }
    else {
        result = withCommonAncestor();
    }
    if (!result) {
        throw new Error("No common ancestor found");
    }
    let { context: c, affectedFrom, affectedTo } = result;
    if (c.events.children.length) {
        splice(root, affected.map(({ path }) => path), c.events.children);
    }
    let lastToBeRelativeTo;
    const relativeContext = new ParsingContext(now, previousParse.cache, (_c) => {
        if (_c.tail) {
            return _c.tail;
        }
        if (lastToBeRelativeTo) {
            if (isEvent(lastToBeRelativeTo)) {
                return lastToBeRelativeTo;
            }
            throw new Error("Last unaffected eventy to relate to is not an event");
        }
    });
    relativeContext.header = { ...previousParse.header };
    relativeContext.ids = { ...previousParse.ids };
    relativeContext.getById = (id) => {
        const path = previousParse.ids[id];
        if (!path) {
            return;
        }
        return get(previousParse.events, path);
    };
    const m = (i) => _change.mapPos(i);
    const mapRange = (r) => {
        return {
            ...r,
            from: m(r.from),
            to: m(r.to),
        };
    };
    const from = affected.at(-1);
    if (from) {
        outer: for (const { eventy, path } of iterateTreeFromPath(root, from.path)) {
            for (const { eventy: newEventy } of iter(c.events)) {
                if (!newEventy.textRanges) {
                    // The root doesn't have text ranges... it probably should though??
                    continue;
                }
                if (isEvent(newEventy)) {
                    lastToBeRelativeTo = newEventy;
                }
                if (newEventy.textRanges.whole.from === eventy.textRanges.whole.from &&
                    newEventy.textRanges.whole.to === eventy.textRanges.whole.to) {
                    if (isEvent(eventy) && eventy.isRelative) {
                        const priorEventPath = getPreviousEventPath(root, path);
                        if (priorEventPath && eventy.fromRelativeTo) {
                            eventy.fromRelativeTo = {
                                ...eventy.fromRelativeTo,
                                path: [...priorEventPath],
                            };
                        }
                        if (priorEventPath && eventy.toRelativeTo) {
                            eventy.toRelativeTo = {
                                ...eventy.toRelativeTo,
                                path: [...priorEventPath],
                            };
                        }
                    }
                    continue outer;
                }
            }
            eventy.textRanges.whole = mapRange(eventy.textRanges.whole);
            if (eventy.textRanges.properties) {
                eventy.textRanges.properties = mapRange(eventy.textRanges.properties);
            }
            if (isGroup(eventy)) {
                // Groups (sections) have a definition range for their header line
                if (eventy.textRanges.definition) {
                    eventy.textRanges.definition = mapRange(eventy.textRanges.definition);
                }
            }
            else {
                eventy.textRanges.datePart = mapRange(eventy.textRanges.datePart);
                eventy.textRanges.definition = mapRange(eventy.textRanges.definition);
                if (eventy.textRanges.recurrence) {
                    eventy.textRanges.recurrence = mapRange(eventy.textRanges.recurrence);
                }
                if (eventy.isRelative) {
                    const priorEventPath = getPreviousEventPath(root, path);
                    if (priorEventPath) {
                        const priorEvent = get(root, priorEventPath);
                        if (priorEvent && isEvent(priorEvent)) {
                            relativeContext.currentPath = [...priorEventPath];
                            relativeContext.tail = priorEvent;
                        }
                        else {
                            relativeContext.currentPath = [...priorEventPath];
                            relativeContext.tail = undefined;
                        }
                    }
                    else {
                        relativeContext.currentPath = [];
                        relativeContext.tail = undefined;
                    }
                    const { from, lines, lengths } = linesAndLengths(newText, _change, [eventy], false);
                    const c = parsePastHeader(from, relativeContext, Array(from).concat(lines), Array(from).concat(lengths));
                    if (c.events.children.length !== 1) {
                        throw new Error("Tried to update relative event and failed");
                    }
                    else {
                        splice(root, [path], c.events.children);
                    }
                }
            }
        }
    }
    mapRanges(previousParse, c, affectedFrom, affectedTo, _change, previousText, newText);
}
function areSiblings(affected) {
    if (affected.length < 2) {
        return true;
    }
    const parent = affected[0].slice(0, -1).join(",");
    const length = affected[0].length;
    const leaves = [affected[0].at(-1)];
    for (let i = 1; i < affected.length; i++) {
        const path = affected[i];
        if (path.length !== length || path.slice(0, -1).join(",") !== parent) {
            return false;
        }
        const leaf = path.at(-1);
        if (!leaves.some((x) => Math.abs(leaf - x) === 1)) {
            return false;
        }
        leaves.push(leaf);
    }
    return true;
}
function mapRanges(previousParse, c, affectedFrom, affectedTo, _change, previousText, newText) {
    const m = (i) => _change.mapPos(i);
    const { foldables } = previousParse;
    let newFoldables = {};
    // all this foldable shit sucks
    for (const [foldableIndex, foldable] of Object.entries(foldables)) {
        const index = m(parseInt(foldableIndex));
        if (index < affectedFrom || index >= affectedTo) {
            newFoldables[index] = {
                ...foldable,
                endIndex: _change.mapPos(foldable.endIndex),
                foldStartIndex: foldable.foldStartIndex !== undefined
                    ? _change.mapPos(foldable.foldStartIndex)
                    : undefined,
                startIndex: foldable.startIndex !== undefined
                    ? _change.mapPos(foldable.startIndex)
                    : undefined,
                startLine: newText.lineAt(_change.mapPos(previousText.line(foldable.startLine + 1).from)).number - 1,
            };
        }
    }
    newFoldables = {
        ...newFoldables,
        ...c.foldables,
    };
    previousParse.foldables = newFoldables;
    let newRanges = [];
    for (const range of previousParse.ranges) {
        const to = m(range.to);
        const from = m(range.from);
        if (to < affectedFrom) {
            newRanges.push(range);
        }
        else if (from >= affectedTo) {
            newRanges.push({
                ...range,
                from,
                to,
            });
        }
    }
    previousParse.ranges = [
        ...newRanges,
        ...c.ranges.filter(({ from, to }) => from >= affectedFrom && to <= affectedTo),
    ].sort(({ from: fromA }, { from: fromB }) => fromA - fromB);
    let newMessages = [];
    for (const message of previousParse.parseMessages) {
        const from = m(message.pos[0]);
        const to = m(message.pos[1]);
        if (to < affectedFrom) {
            newMessages.push(message);
        }
        else if (from > affectedTo) {
            newMessages.push({
                ...message,
                pos: [from, to],
            });
        }
    }
    previousParse.parseMessages = [...newMessages, ...c.parseMessages].sort(({ pos: posA }, { pos: posB }) => posA[0] - posB[0]);
}
function splice(root, affectedPaths, eventies) {
    if (!affectedPaths.length) {
        root.children.splice(root.children.length, 0, ...eventies);
        return;
    }
    if (!areSiblings(affectedPaths)) {
        throw new Error("Can't splice non-siblings");
    }
    const youngest = affectedPaths.reduce((prev, curr) => {
        return curr.at(-1) < prev.at(-1) ? curr : prev;
    });
    while (youngest.length > 1) {
        root = root.children[youngest.shift()];
    }
    root.children.splice(youngest[0], affectedPaths.length, ...eventies);
}
function mapParseThroughChanges(parse, changes, previousText, now) {
    const changesArray = asArray(changes);
    if (!changesArray.length) {
        parse.parser.incremental = true;
        return parse;
    }
    let text = previousText;
    let offset = 0;
    for (let i = 0; i < changesArray.length; i++) {
        const { fromA, toA, inserted } = changesArray[i];
        const adjustedFrom = fromA + offset;
        const adjustedTo = toA + offset;
        const adjustedChange = {
            fromA: adjustedFrom,
            toA: adjustedTo,
            fromB: adjustedFrom,
            toB: adjustedFrom + inserted.length,
            inserted,
        };
        let eventyIterator = iter(parse.events);
        let { done, eventy, path } = next(eventyIterator);
        if (done || !eventy.children.length) {
            throw new Error("unimplemented");
        }
        ({ done, eventy, path } = next(eventyIterator));
        if (done || !eventy || !path) {
            throw new Error("Nothing to graft");
        }
        graft({
            changedRange: adjustedChange,
            previousParse: parse,
            now,
            done,
            eventy,
            eventyIterator,
            previousText: text,
            path,
        });
        text = text.replace(adjustedFrom, adjustedTo, inserted);
        offset += inserted.length - (toA - fromA);
    }
    parse.parser.incremental = true;
    return parse;
}
export function incrementalParse(previousText, changes = ChangeSet.empty(previousText.length), previousParse, now) {
    const text = () => previousText instanceof Text
        ? previousText
        : Text.of(Array.isArray(previousText) ? previousText : previousText.split("\n"));
    const bail = () => {
        return parse(changes.apply(text()), previousParse?.cache, now);
    };
    if (changes.length !== previousText.length) {
        throw new RangeError("Applying change set to a document with the wrong length");
    }
    if (!previousParse || !previousParse.events.children.length) {
        return bail();
    }
    if (changes.empty) {
        return previousParse ? previousParse : bail();
    }
    const _previousParse = { ...previousParse };
    try {
        return mapParseThroughChanges(_previousParse, changes, text(), now);
    }
    catch (e) {
        if (e.message === "Won't reparse over ided events") {
            return bail();
        }
        if (e.message === "Can't reparse in header area") {
            return bail();
        }
        console.error(e);
        return bail();
    }
}
