// @markwhen/parser@1.0.1/lib/ParsingContext.js
import { DateTime, SystemZone } from "luxon";
import { EventGroup, push, get, toDateRange, } from "./Types.js";
import { parseZone } from "./zones/parseZone.js";
export class ParsingContext {
    constructor(now, cache, getPriorEvent) {
        this.parseMessages = [];
        this.documentMessages = [];
        this.sectionLevels = [];
        this.events = new EventGroup();
        this.ids = {};
        this.paletteIndex = 0;
        this.earliest = undefined;
        this.latest = undefined;
        this.maxDuration = undefined;
        this.currentPath = [];
        this.foldables = {};
        this.foldableSections = [];
        this.ranges = [];
        this.header = {};
        this.cache = cache;
        if (typeof now === "string") {
            const parsed = DateTime.fromISO(now);
            if (parsed.isValid) {
                this.now = parsed;
            }
            else {
                this.now = DateTime.now();
            }
        }
        else if (!now) {
            this.now = DateTime.now();
        }
        else {
            this.now = now;
        }
        if (getPriorEvent) {
            this.priorEvent = () => getPriorEvent(this);
        }
    }
    get zonedNow() {
        return this.now.setZone(this.timezone);
    }
    get timezone() {
        return this.parentZone();
    }
    currentFoldableSection() {
        return this.foldableSections[this.foldableSections.length - 1];
    }
    currentFoldableComment() {
        return this.foldables["comment"];
    }
    startFoldableSection(f) {
        this.foldableSections.push(f);
    }
    startFoldable(f) {
        this.foldables[f.type] = f;
    }
    finishFoldableSection(line, endIndex) {
        const currentFoldableSection = this.foldableSections.pop();
        if (currentFoldableSection) {
            if (currentFoldableSection.startLine < line - 1) {
                this.foldables[currentFoldableSection.startIndex] = {
                    ...currentFoldableSection,
                    endIndex,
                };
            }
        }
    }
    finishFoldableComment(lineNumber) {
        const commentFoldable = this.currentFoldableComment();
        if (commentFoldable) {
            if (commentFoldable.startLine < lineNumber - 1) {
                // We had had a foldable comment section that we can close off, since this line
                // is not a comment.
                this.foldables[commentFoldable.startIndex] = {
                    ...commentFoldable,
                };
            }
            delete this.foldables["comment"];
        }
    }
    push(node) {
        const { path, tail: newTail } = push(node, this.events, this.currentPath.slice(0, -1), this.tail);
        if (newTail) {
            if (!this.head) {
                this.head = newTail;
            }
            this.tail = newTail;
        }
        this.currentPath = path;
        return path;
    }
    endCurrentGroup(to, lineTo) {
        // Assign text range to the group being closed (before popping)
        // currentPath has an extra index for where children would go, so we need to
        // slice off the last element to get the actual group path
        const groupPath = this.currentPath.slice(0, -1);
        const group = get(this.events, groupPath);
        if (group && group.textRanges) {
            group.textRanges.whole = {
                ...group.textRanges.whole,
                to: to,
            };
        }
        this.currentPath.pop();
        this.finishFoldableSection(lineTo.line, to);
    }
    toTimeline() {
        const maxDurationDays = this.maxDuration
            ? this.maxDuration / 1000 / 60 / 60 / 24
            : 0;
        return {
            events: this.events,
            ids: this.ids,
            ranges: this.ranges.sort(({ from: fromA }, { from: fromB }) => fromA - fromB),
            foldables: this.foldables,
            header: this.header,
            parseMessages: this.parseMessages.sort(({ pos: posA, message: messageA }, { pos: posB, message: messageB }) => posA[0] - posB[0]),
            documentMessages: this.documentMessages,
        };
    }
    priorEvent() {
        return this.tail;
    }
    parentZone() {
        const p = [...this.currentPath];
        p.pop();
        while (p.length) {
            const parent = get(this.events, p);
            const zone = timezoneFromProperties(parent?.properties ?? [], this.cache);
            if (zone) {
                return zone;
            }
            p.pop();
        }
        return (parseZone(this.header.timezone ?? this.header.tz, this.cache) ??
            SystemZone.instance);
    }
    priorEventToDateTime() {
        const prior = this.priorEvent();
        if (!prior) {
            return;
        }
        return toDateRange(prior.dateRangeIso).toDateTime;
    }
    priorEventFromDateTime() {
        const prior = this.priorEvent();
        if (!prior) {
            return;
        }
        return toDateRange(prior.dateRangeIso).fromDateTime;
    }
    getById(id) {
        const path = this.ids[id];
        if (!path) {
            return;
        }
        return get(this.events, path);
    }
}
export function timezoneFromProperties(properties, cache) {
    if (Array.isArray(properties)) {
        const timezoneProperty = properties.find(([k, v]) => {
            return ((k === "tz" || k === "timezone") &&
                (typeof v === "string" || typeof v === "number"));
        });
        if (timezoneProperty) {
            return parseZone(timezoneProperty[1], cache);
        }
    }
    else {
        const tz = properties?.timezone ?? properties?.tz;
        if (tz) {
            return parseZone(tz, cache);
        }
    }
}
