// @markwhen/parser@1.0.1/lib/utilities/dateRangeToString.js
export var Weight;
(function (Weight) {
    Weight[Weight["SECOND"] = 0] = "SECOND";
    Weight[Weight["QUARTER_MINUTE"] = 1] = "QUARTER_MINUTE";
    Weight[Weight["MINUTE"] = 2] = "MINUTE";
    Weight[Weight["QUARTER_HOUR"] = 3] = "QUARTER_HOUR";
    Weight[Weight["HOUR"] = 4] = "HOUR";
    Weight[Weight["DAY"] = 5] = "DAY";
    Weight[Weight["MONTH"] = 6] = "MONTH";
    Weight[Weight["YEAR"] = 7] = "YEAR";
    Weight[Weight["DECADE"] = 8] = "DECADE";
    Weight[Weight["CENT"] = 9] = "CENT";
})(Weight || (Weight = {}));
export function dateScale(dateTime) {
    if (dateTime.second === 0) {
        if (dateTime.minute === 0) {
            if (dateTime.hour === 0) {
                if (dateTime.day === 1) {
                    if (dateTime.month === 1) {
                        if (dateTime.year % 100 === 0) {
                            return Weight.CENT;
                        }
                        if (dateTime.year % 10 === 0) {
                            return Weight.DECADE;
                        }
                        return Weight.YEAR;
                    }
                    return Weight.MONTH;
                }
                return Weight.DAY;
            }
            return Weight.HOUR;
        }
        else if (dateTime.minute % 15 == 0) {
            return Weight.QUARTER_HOUR;
        }
        return Weight.MINUTE;
    }
    else if (dateTime.second % 15 === 0) {
        return Weight.QUARTER_MINUTE;
    }
    return Weight.SECOND;
}
export const muricaMap = {
    [Weight.CENT]: {
        format: "yyyy",
        separator: " - ",
    },
    [Weight.DECADE]: {
        format: "yyyy",
        separator: " - ",
    },
    [Weight.YEAR]: {
        format: "yyyy",
        separator: " - ",
    },
    [Weight.MONTH]: {
        format: "MMM yyyy",
        separator: " - ",
    },
    [Weight.DAY]: {
        format: "MMM d yyyy",
        separator: " - ",
    },
    [Weight.HOUR]: {
        format: `MMM d yyyy h:mma`,
        separator: " - ",
    },
    [Weight.MINUTE]: {
        format: `MMM d yyyy h:mma`,
        separator: " - ",
    },
    [Weight.QUARTER_HOUR]: {
        format: `MMM d yyyy h:mma`,
        separator: " - ",
    },
    [Weight.QUARTER_MINUTE]: {
        format: `MMM d yyyy h:mma`,
        separator: " - ",
    },
    [Weight.SECOND]: {
        format: `yyyy-MM-ddThh:mm:ssZ`,
        separator: " - ",
    },
};
export const ISOMap = {
    [Weight.CENT]: {
        format: "yyyy",
        separator: "/",
    },
    [Weight.DECADE]: {
        format: "yyyy",
        separator: "/",
    },
    [Weight.YEAR]: {
        format: "yyyy",
        separator: "/",
    },
    [Weight.MONTH]: {
        format: "yyyy-MM",
        separator: "/",
    },
    [Weight.DAY]: {
        format: "yyyy-MM-dd",
        separator: "/",
    },
    [Weight.HOUR]: {
        format: `d MMM yyyy T`,
        separator: " - ",
    },
    [Weight.QUARTER_HOUR]: {
        format: `d MMM yyyy T`,
        separator: " - ",
    },
    [Weight.MINUTE]: {
        format: `d MMM yyyy T`,
        separator: " - ",
    },
    [Weight.QUARTER_MINUTE]: {
        format: `d MMM yyyy T`,
        separator: " - ",
    },
    [Weight.SECOND]: {
        format: `yyyy-MM-ddThh:mm:ssZ`,
        separator: " - ",
    },
};
export function dateRangeToString({ fromDateTime, toDateTime, }, formap = ISOMap) {
    const commonScale = Math.min(dateScale(fromDateTime), dateScale(toDateTime));
    const format = formap[commonScale];
    return `${fromDateTime.toFormat(format.format)}${format.separator || " - "}${toDateTime.toFormat(format.format)}`;
}
